### DISCUSSION TITLE
How to allow a client in a new realm have permissions for external-internal token-exchange using new UI

### DISCUSSION THREAD

**Discussion by outofthisworld** [keycloak/keycloak#19880] [Q&A]
I can't for the life of me figure out how to enable the permission of token-exchange for a client.

Everything is working great, apart from the error I'm getting: Client not allowed to exchange

I'm using the new UI and a first time user of keycloak - it's not as intuitive as I would have hoped. 

I have found documentation on the keycloak website. However, my client doesn't have the 'permissions' tab. Permissions is under authorization - and it's quite different from the docs

<img width="1122" alt="Screen Shot 2023-04-22 at 10 59 39 AM" src="https://user-images.githubusercontent.com/8555871/233746511-a81adc19-245f-4b15-bd80-6d0c63e7436d.png">

**Comment by outofthisworld** [comment]
@jonkoops

**Reply by jonkoops** [reply]
I can't reproduce this without detailed steps. Please report this as an issue with the steps to reproduce it.

**Reply by outofthisworld** [reply]
@jonkoops I'm more just asking for instructions right now, not sure if it's an issue. I think some of the documentation screenshots are outdated since the UI updates. [See Keycloak Docs Here](https://www.keycloak.org/docs/latest/securing_apps/#_token-exchange).

Trying to follow this guide and enable token exchange permissions for a client in a newly created realm - but I don't have half the stuff shown in the docs when using the latest version of Keycloak

After I enable `Authorization` and `Authentication` on the client, I do not get access to the `Permissions` tab as shown in the docs. I believe this is because the updates may have changed how this is now accomplished.

**Reply by outofthisworld** [reply]
Bump

**Reply by jonkoops** [reply]
Sorry I do not have time at the moment to look into this, if you feel this is a problem with the documentation feel free to submit an issue.

**Comment by rcarpa** [comment]
In case somebody else gets directed to this discussion by search: everything actually works as instructed in the documentation, but you have to enable the `admin-fine-grained-authz` preview feature on top of the `token-exchange` one for the relevant option to appear in the webUI

**Reply by Crystark** [reply]
Indeed, thanks for the pointer. I was convinced it was the application which had evolved as that Permission tab is elsewhere.
But again maybe I should have just read the documentation from the beginning 😅
https://www.keycloak.org/docs/latest/securing_apps/#_token-exchange

**Reply by ffroliva** [reply]
I think enabling token exchange is quite complicated and involves multiple steps. I think there should be a way to just (given two internal clients from a given realm) configure sensible default permission between them to enable token-exchange. an API for that would be interesting.

**Comment by nabim777** [comment]
> I can't for the life of me figure out how to enable the permission of token-exchange for a client.
> 
> Everything is working great, apart from the error I'm getting: Client not allowed to exchange
> 
> I'm using the new UI and a first time user of keycloak - it's not as intuitive as I would have hoped.
> 
> I have found documentation on the keycloak website. However, my client doesn't have the 'permissions' tab. Permissions is under authorization - and it's quite different from the docs
> 
> <img alt="Screen Shot 2023-04-22 at 10 59 39 AM" width="1122" src="https://user-images.githubusercontent.com/8555871/233746511-a81adc19-245f-4b15-bd80-6d0c63e7436d.png">

 In order to have the 'permissions' tab in the client, we have to enable it with the environment variable. 

If you are running locally then we can enable by running with extra env
```bash
bin/kc.[sh|bat] build --features="preview"
```
 ## or

If you are running keycloak using docker, then we can enable by adding below env:
```console
 KC_FEATURES=preview
 ```

After this we get this `permission` tab on client as shown in image below:
![image](https://github.com/user-attachments/assets/2267bc4e-8b29-446a-b7f3-2f21749ae3a7)


![image](https://github.com/user-attachments/assets/f683f047-4c4f-4f72-bc46-3b670fcc4eb9)


for more [here](https://www.keycloak.org/server/features)
